bgpd/bgp_route.c

```diff
@@ -5159,6 +5159,17 @@ void bgp_update(struct peer *peer, const struct prefix *p, uint32_t addpath_id,
 		}
 
 	new_attr = *attr;
+	if (CHECK_FLAG(peer->af_flags[afi][safi], PEER_FLAG_SOFT_RECONFIG)) {
+		/* For soft-reconfiguration, the attributes are interned. If
+		 * the EVPN attributes are set, they are ref-counted. We
+		 * perform a shallow-copy of the attribtues above, thus need to
+		 * increase the EVPN refcount to avoid a double-free later on.
+		 */
+		struct bgp_route_evpn *bre = bgp_attr_get_evpn_overlay(attr);
+		if (bre && bre->refcnt)
+			bre->refcnt++;
```

Comment:
This should be done by `bgp_attr_intern()`, isn't?

Reply:
I indeed missed the `bgp_attr_intern` further down the line, which confuses me even more. All exit paths use `bgp_attr_flush`, which properly uses the refcounts to free the EVPN attribute.

Now I don't understand why this patch works, but empirically it does. A few updates are enough to corrupt memory in various places (hit malloc today instead of what I put in the commit), while the patch makes it work for a few thousand updates (tested).

I'm hunting this one deeper. In the meantime, I dropped the soft-reconfig related commit from the PR until I get to the bottom of this. At least this way EVPN attributes will work again :)